### PR TITLE
Refactor part of ValueTuple into a new Hash class, update hashing algorithm

### PR DESCRIPTION
**Pull request by jamesqo**
Made as a result of the discussion in #8193; basically, `ValueTuple` has a bunch of hashing logic that could probably be separated out into a new file/be used in other places in the repo. I've added a new `Hash` class in `Common` based off [my previous proposal](https://github.com/dotnet/corefx/issues/8034) and refactored the class to use that. (Feel free to suggest if I should change the name to something like `HashCodeHelper`/etc, since this is for internal use.)

On a related tangent, I also changed the hash code algorithm as per @VSadov's suggestion; the new formula to combine hash codes is now

```cs
public static int Combine(int h1, int h2)
{
    return ((h1 << 5) + h1 + (h1 >> 27)) ^ h2;
}
```

This reduces the likelihood of collisions if the bottom 27 bits of `h1` are 0, since the new formula takes all bits into account equally.

Other changes:

- I've changed the overloads of `CombineHashCodes` (now `Combine`) that take 3 or more parameters to just combine the last hash with others' combined hash code. This should allow the hash to spread more for lower hash codes; for example, if you wrote

```cs
var random = new Random();

var r1 = random.Next(1);
var r2 = random.Next(1);
var r3 = random.Next(1);
var r4 = random.Next(1);

var tuple = ValueTuple.Create(r1, r2, r3, r4);
int hashCode = tuple.GetHashCode();
Console.WriteLine($"0x{hashCode:X}");
```

In the [previous implementation](https://github.com/dotnet/corefx/blob/master/src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs#L258), the number could only take up a maximum of 11 bits assuming the first random number was `1`. With the new one, it simply keeps shifting each hash left so it can take up to 16 bits.

- Removed the calls to `EqualityComparer<T>.Default.GetHashCode`, and replaced it with a custom implementation that checks for null/calls `GetHashCode`.

cc: @jcouv @VSadov @stephentoub

### CODE REVIEW

**Comment by hughbe** [comment]
Interesting! I'd love to see this become public API at some poiint

File: src/Common/src/System/Numerics/Hashing/Hash.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using System;
+
+namespace System.Numerics.Hashing
+{
+    internal static class Hash
+    {
+        public static int Combine(int h1, int h2)
+        {
+            return ((h1 << 5) + h1 + (h1 >> 27)) ^ h2;
+        }
```
Comment (VSadov) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: This code looks a lot like   ``` (ROL(h1, 5) + h1) ^ h2 ``` , but not quite that. It could be beneficial to match the ROL pattern since JIT may generate better code for it. - something like:
``` (((h1 << 5) | ((uint)h1 >> 27)) + h1) ^ h2 ``` 

(see: https://github.com/dotnet/coreclr/pull/1830).

File: src/Common/src/System/Numerics/Hashing/Hash.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                GetHashCode(o6), GetHashCode(o7), GetHashCode(o8));
+        }
+        
+        private static int GetHashCode<T>(T obj)
+        {
+            // Emulate the behavior of EqualityComparer<T>.Default,
+            // without introducing a virtual method call.
+            // For structs, it looks like the JIT will eliminate
+            // the null check here and simply call obj.GetHashCode().
+            return obj == null ? 0 : obj.GetHashCode();
```
Comment (omariom) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: I suspect it will box the value if `T` is a enum.

Comment (jamesqo) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: @omariom You appear to be right, the value is getting boxed: https://gist.github.com/anonymous/9175d70ec5e9c02b3877d6e63e591591

I'll change it back to use `EqualityComparer` for now. (coreclr should probably handle this better, though.)

File: src/Common/src/System/Numerics/Hashing/Hash.cs (hunk 1 above)
Comment (jamesqo) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: @VSadov Actually, it seems the current way it's written generates more efficient code. I did a disassembly dump of [this program](https://gist.github.com/jamesqo/73e8018b05c02ec3310953f45415a364) and here were the results: https://gist.github.com/7285fdea2585232610a367fd438819a9

**Comment by jamesqo** [comment]
Thanks everyone for the review! I've responded to PR feedback.

**Comment by VSadov** [comment]
@stephentoub @jcouv

**Comment by VSadov** [comment]
LGTM

File: src/Common/src/System/Numerics/Hashing/Hash.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Numerics.Hashing
+{
+    internal static class Hash
```
Comment (bartonjs) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: Certainly my first thought at the bareword "hash" is a cryptographic hash/digest.  But maybe I'm just tainted by being a specialist.

File: src/Common/src/System/Numerics/Hashing/Hash.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        public static int Combine(int h1, int h2)
+        {
+            // The jit optimizes this to use the ROL instruction on x86
+            // Related GitHub pull request: dotnet/coreclr#1830
+            uint shift5 = ((uint)h1 << 5) | ((uint)h1 >> 27);
+            return ((int)shift5 + h1) ^ h2;
```
Comment (stephentoub) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: @jamesqo, where else are you expecting this is going to be used?  I can see the value in putting this one function in a shared location, but why all of the overloads that take 3-N arguments?  Those seem like they're fairly unique to a library as for tuples and would be better served just being in the tuples library as they previously were.  As you've currently factored this, any library that wants to use this function is going to need to pull in all of the rest.  Then as for this one function, I know @VSadov suggested this as an alternative to help avoid collisions... do we know that it's a net-gain performance-wise?  It may help to reduce collisions, but it would also seem to be a more complicated/expensive operation (your comment mentions it's optimized... does that make it faster than the previous hash code implementation, or just faster than the new code would have been without the optimization?)

Comment (jamesqo) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: > I can see the value in putting this one function in a shared location, but why all of the overloads that take 3-N arguments?

Classes that have 3+ fields, e.g. could be useful [here](https://github.com/dotnet/corefx/pull/8163/files#diff-b6b7e77633e5d7410bd990dd90124e3fR742) in `Cookie`. I don't expect the other methods that aren't used will contribute to much of the code size of the outputted IL; if you want I can have maybe up to 3 overloads of `Combine` in the shared file, and declare the rest in `ValueTuple`.

> do we know that it's a net-gain performance-wise? It may help to reduce collisions, but it would also seem to be a more complicated/expensive operation

It isn't faster than the previous implementation, but I certainly don't think it would be much slower. If I remember correctly the number of registers/instructions used in the disassembly was the same-- before it was shl + add + xor, now it's rol + add + xor. If there was any slowdown, I think the overhead of all the virtual method calls to `EqualityComparer<T>.Default.GetHashCode` would overshadow that here.

Comment (stephentoub) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: If @VSadov and @jcouv are happy with the revised hash function, then that could be changed in-place in ValueTuple.  Separating out all of these methods into their own internal common class just doesn't seem worth it to me; if all of them are there, then any code which wants to include just one will need to compile in them all (which affects IL size, makes it more difficult to understand what code is dead, etc.), and if only some are here, then ValueTuple will be calling to the common class for some of the combinations and doing some itself, which is in and of itself confusing.  If at some point we wanted to expose a public class for this, then it may make sense, though even then it's not clear to me we'd want all of these overloads.  In the meantime, I'd be ok with having the int,int hash combining function itself in a common class, so that various types can use the same hash function, but everything else seems superfluous to me.

Comment (jamesqo) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: Your concerns should be addressed: I moved the `Hash` class into the ValueTuple assembly, and added a new class in Common (`HashHelpers`) that contains the actual algorithm for the hash, and now `Hash` just forwards to that.

File: src/System.ValueTuple/src/System/Numerics/Hashing/Hash.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace System.Numerics.Hashing
+{
+    internal static class Hash
```
Comment (stephentoub) [inline src/System.ValueTuple/src/System/Numerics/Hashing/Hash.cs]: Why is this class valuable here? Why not just delete this and revert all of the changes to ValueTuple, other than using the new HashHelpers.Combine?  I'm not seeing what the value is (no pun intended :))

Comment (jamesqo) [inline src/System.ValueTuple/src/System/Numerics/Hashing/Hash.cs]: I kept it since it saves on lines of code in `ValueTuple`; if we didn't have these functions we'd have to write something like

```cs
int result = EqualityComparer<T1>.Default.GetHashCode(Item1);
result = HashHelpers.Combine(result, EqualityComparer<T2>.Default.GetHashCode(Item2));
result = HashHelpers.Combine(result, EqualityComparer<T3>.Default.GetHashCode(Item3));
...
return result;
```

which is significantly longer than just `Hash.CombineHashCodes(Item1, Item2, Item3, ...)`.

Comment (stephentoub) [inline src/System.ValueTuple/src/System/Numerics/Hashing/Hash.cs]: > if we didn't have these functions we'd have to write something like

No, we already have the CombineHashCodes methods in ValueTuple.

> which is significantly longer

But it also introduces another generic method that'll need to be compiled, and will likely need a new compilation of it for every combination of Ts that come in (assuming they're not all reference types).  Plus every item is getting passed to another method, so if any of them are big, it's yet another copy.

Comment (stephentoub) [inline src/System.ValueTuple/src/System/Numerics/Hashing/Hash.cs]: I was thinking this whole PR would boil down to just having the new HashHelpers.Combine(int, int) method in its own file, and either deleting the existing ValueTuple.CombineHashCodes(int, int) method and replacing its call sites with calls to HashHelpers.Combine(int, int), or simply having ValueTuple.CombineHashCodes(int, int) call HashHelpers.Combine(int, int).

Comment (jamesqo) [inline src/System.ValueTuple/src/System/Numerics/Hashing/Hash.cs]: @stephentoub Ok; I re-added `ValueTuple.CombineHashCodes` and had it forward to the helper class. The diff is now only ~60 LOC.

File: src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs [hunk 6]
```diff
@@ -255,7 +256,7 @@ internal static int CombineHashCodes(int h1, int h2, int h3)
 
         internal static int CombineHashCodes(int h1, int h2, int h3, int h4)
         {
-            return CombineHashCodes(CombineHashCodes(h1, h2), CombineHashCodes(h3, h4));
+            return CombineHashCodes(CombineHashCodes(h1, h2, h3), h4);
```
Comment (stephentoub) [inline src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: Why did you change these calls?  The previous version had three calls to CombineHashCodes total.  This one has four: two in this method, and then two in the CombineHashCodes(int, int, int) overload.  Same question about the other changes below.  What's the benefit of these changes?

Comment (jamesqo) [inline src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: @stephentoub This is actually just 3, since the 3-overload `CombineHashCodes` method (which will probably be inlined) is just calling the 2-overload one (which does the actual work) twice, and this calls the 2-overload one once. I mentioned why I changed it in my post: it 'spreads' the hash farther up for low numbers.

e.g. if you have a tuple of `1 2 3 4`, then in the previous implementation you would take the hash of 1 and 2, and since all the upper bits are 0 the furthest bit that would get affected is maybe the 6th bit. Then you hash 3 and 4 together, and the furthest bit that gets affected is the 7th bit. Then when you combine them again, the furthest bit that could be set is 11th, meaning you wasted ~21 of them.

With the new implementation it keeps spreading, so you would take 1 and 2, hash them to get 6 bits, hash it with 3 to get 11 bits, then with 4 to get 16 bits. Basically less bits are wasted, meaning a lower likelihood of collisions.

**Comment by jamesqo** [comment]
Test Innerloop OSX Debug Build and Test

File: src/Common/src/System/Numerics/Hashing/HashHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
```
Comment (justinvp) [inline src/Common/src/System/Numerics/Hashing/HashHelpers.cs]: Nit: Are these namespace imports needed?

File: src/Common/src/System/Numerics/Hashing/HashHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public static int Combine(int h1, int h2)
+        {
+            // The jit optimizes this to use the ROL instruction on x86
+            // Related GitHub pull request: dotnet/coreclr#1830
+            uint shift5 = ((uint)h1 << 5) | ((uint)h1 >> 27);
+            return ((int)shift5 + h1) ^ h2;
+        }
+    }
+}
```
Comment (justinvp) [inline src/Common/src/System/Numerics/Hashing/HashHelpers.cs]: Nit: missing newline at the end of the file

File: src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs [hunk 9]
```diff
@@ -242,10 +243,10 @@ string ITupleInternal.ToStringEnd()
         public static ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>> Create<T1, T2, T3, T4, T5, T6, T7, T8>(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5, T6 item6, T7 item7, T8 item8) =>
             new ValueTuple<T1, T2, T3, T4, T5, T6, T7, ValueTuple<T8>>(item1, item2, item3, item4, item5, item6, item7, ValueTuple.Create(item8));
 
-        // From System.Web.Util.HashCodeCombiner
         internal static int CombineHashCodes(int h1, int h2)
         {
-            return (((h1 << 5) + h1) ^ h2);
+            // Forward to helper class in Common for this
```
Comment (justinvp) [inline src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: Nit: Is this comment necessary?

Comment (jamesqo) [inline src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: I added more to the comment, explaining exactly why the logic was separated out into Common.

**Comment by jamesqo** [comment]
@jcouv What's your opinion on this? Are you OK with the hashing algorithm? (This is blocking a few other places from taking advantage of the updated/optimized algorithm.)

**Comment by jcouv** [comment]
LGTM

**Comment by jamesqo** [comment]
Thank you for merging this @VSadov! I'm going to make the corresponding changes to Tuple in a week when I'm back at my computer.

**Comment by jcouv** [comment]
I forgot to ping @dsyme and @KevinRansom as FYI for this change. 
The hash code method for `ValueTuple` is getting an upgrade.
The corresponding method for `System.Tuple` will also be upgraded at some point.

File: src/Common/src/System/Numerics/Hashing/Hash.cs (hunk 2 above)
Comment (jamesqo) [inline src/Common/src/System/Numerics/Hashing/Hash.cs]: @omariom, this is a few months later, but... is enums the only case this penalizes for? I wonder if it would be a good trade off to box for enums, and avoid a virtual method call for all other cases.
